Add Evaluation / Feedback Examples

vowelparrot opened the pull request.

vowelparrot committed:
[WIP] Add Feedback Crud Draft

jettro committed:
Fixes issue #5072 - adds additional support to Weaviate (#5085)

Implementation is similar to search_distance and where_filter

# adds 'additional' support to Weaviate queries

Co-authored-by: Dev 2049 <dev.dev2049@gmail.com>

eyurtsev committed:
Improve effeciency of TextSplitter.split_documents, iterate once (#5111)

# Improve TextSplitter.split_documents, collect page_content and
metadata in one iteration

## Who can review?

Community members can review the PR once tests pass. Tag
maintainers/contributors who might be interested:

@eyurtsev In the case where documents is a generator that can only be
iterated once making this change is a huge help. Otherwise a silent
issue happens where metadata is empty for all documents when documents
is a generator. So we expand the argument from `List[Document]` to
`Union[Iterable[Document], Sequence[Document]]`

---------

Co-authored-by: Steven Tartakovsky <tartakovsky.developer@gmail.com>

jamie256 committed:
WhyLabs callback (#4906)

# Add a WhyLabs callback handler

* Adds a simple WhyLabsCallbackHandler
* Add required dependencies as optional
* protect against missing modules with imports
* Add docs/ecosystem basic example

based on initial prototype from @andrewelizondo

> this integration gathers privacy preserving telemetry on text with
whylogs and sends stastical profiles to WhyLabs platform to monitoring
these metrics over time. For more information on what WhyLabs is see:
https://whylabs.ai

After you run the notebook (if you have env variables set for the API
Keys, org_id and dataset_id) you get something like this in WhyLabs:
![Screenshot
(443)](https://github.com/hwchase17/langchain/assets/88007022/6bdb3e1c-4243-4ae8-b974-23a8bb12edac)

Co-authored-by: Andre Elizondo <andre@whylabs.ai>
Co-authored-by: Dev 2049 <dev.dev2049@gmail.com>

whiskyboy committed:
Add AzureCognitiveServicesToolkit to call Azure Cognitive Services API (#5012)

# Add AzureCognitiveServicesToolkit to call Azure Cognitive Services
API: achieve some multimodal capabilities

This PR adds a toolkit named AzureCognitiveServicesToolkit which bundles
the following tools:
- AzureCogsImageAnalysisTool: calls Azure Cognitive Services image
analysis API to extract caption, objects, tags, and text from images.
- AzureCogsFormRecognizerTool: calls Azure Cognitive Services form
recognizer API to extract text, tables, and key-value pairs from
documents.
- AzureCogsSpeech2TextTool: calls Azure Cognitive Services speech to
text API to transcribe speech to text.
- AzureCogsText2SpeechTool: calls Azure Cognitive Services text to
speech API to synthesize text to speech.

This toolkit can be used to process image, document, and audio inputs.
---------

Co-authored-by: Dev 2049 <dev.dev2049@gmail.com>

Ayan-Bandyopadhyay committed:
Add link to Psychic from document loaders documentation page (#5115)

# Add link to Psychic from document loaders documentation page

In my previous PR I forgot to update `document_loaders.rst` to link to
`psychic.ipynb` to make it discoverable from the main documentation.

dev2049 committed:
bump 178 (#5130)

amicus-veritatis committed:
docs: fix minor typo + add wikipedia package installation part in human_input_llm.ipynb (#5118)

# Fix typo + add wikipedia package installation part in
human_input_llm.ipynb
This PR
1. Fixes typo ("the the human input LLM"), 
2. Addes wikipedia package installation part (in accordance with
`WikipediaQueryRun`
[documentation](https://python.langchain.com/en/latest/modules/agents/tools/examples/wikipedia.html))

in `human_input_llm.ipynb`
(`docs/modules/models/llms/examples/human_input_llm.ipynb`)

tommasodelorenzo committed:
solving #2887 (#5127)

# Allowing openAI fine-tuned models
Very simple fix that checks whether a openAI `model_name` is a
fine-tuned model when loading `context_size` and when computing call's
cost in the `openai_callback`.

Fixes #2887 
---------

Co-authored-by: Dev 2049 <dev.dev2049@gmail.com>

vowelparrot committed:
Update Notebook with Evaluation Examples

TMRolle committed:
Improve PlanningOutputParser whitespace handling (#5143)

Some LLM's will produce numbered lists with leading whitespace, i.e. in
response to "What is the sum of 2 and 3?":
```
Plan:
  1. Add 2 and 3.
  2. Given the above steps taken, please respond to the users original question.
```
This commit updates the PlanningOutputParser regex to ignore leading
whitespace before the step number, enabling it to correctly parse this
format.

jeffvestal committed:
Add ElasticsearchEmbeddings class for generating embeddings using Elasticsearch models (#3401)

This PR introduces a new module, `elasticsearch_embeddings.py`, which
provides a wrapper around Elasticsearch embedding models. The new
ElasticsearchEmbeddings class allows users to generate embeddings for
documents and query texts using a [model deployed in an Elasticsearch
cluster](https://www.elastic.co/guide/en/machine-learning/current/ml-nlp-model-ref.html#ml-nlp-model-ref-text-embedding).

### Main features:

1. The ElasticsearchEmbeddings class initializes with an Elasticsearch
connection object and a model_id, providing an interface to interact
with the Elasticsearch ML client through
[infer_trained_model](https://elasticsearch-py.readthedocs.io/en/v8.7.0/api.html?highlight=trained%20model%20infer#elasticsearch.client.MlClient.infer_trained_model)
.
2. The `embed_documents()` method generates embeddings for a list of
documents, and the `embed_query()` method generates an embedding for a
single query text.
3. The class supports custom input text field names in case the deployed
model expects a different field name than the default `text_field`.
4. The implementation is compatible with any model deployed in
Elasticsearch that generates embeddings as output.

### Benefits:

1. Simplifies the process of generating embeddings using Elasticsearch
models.
2. Provides a clean and intuitive interface to interact with the
Elasticsearch ML client.
3. Allows users to easily integrate Elasticsearch-generated embeddings.

Related issue https://github.com/hwchase17/langchain/issues/3400

---------

Co-authored-by: Dev 2049 <dev.dev2049@gmail.com>

0xcaffeinated committed:
Adding Weather Loader (#5056)



Co-authored-by: Tyler Hutcherson <tyler.hutcherson@redis.com>
Co-authored-by: Dev 2049 <dev.dev2049@gmail.com>

dakinggg committed:
Add MosaicML inference endpoints (#4607)

# Add MosaicML inference endpoints
This PR adds support in langchain for MosaicML inference endpoints. We
both serve a select few open source models, and allow customers to
deploy their own models using our inference service. Docs are here
(https://docs.mosaicml.com/en/latest/inference.html), and sign up form
is here (https://forms.mosaicml.com/demo?utm_source=langchain). I'm not
intimately familiar with the details of langchain, or the contribution
process, so please let me know if there is anything that needs fixing or
this is the wrong way to submit a new integration, thanks!

I'm also not sure what the procedure is for integration tests. I have
tested locally with my api key.

## Who can review?
@hwchase17

---------

Co-authored-by: Harrison Chase <hw.chase.17@gmail.com>

edwardzjl committed:
Empty check before pop (#4929)

# Check whether 'other' is empty before popping

This PR could fix a potential 'popping empty set' error.

Co-authored-by: Junlin Zhou <jlzhou@zjuici.com>

jlowin committed:
Add async versions of predict() and predict_messages() (#4867)

# Add async versions of predict() and predict_messages()

#4615 introduced a unifying interface for "base" and "chat" LLM models
via the new `predict()` and `predict_messages()` methods that allow both
types of models to operate on string and message-based inputs,
respectively.

This PR adds async versions of the same (`apredict()` and
`apredict_messages()`) that are identical except for their use of
`agenerate()` in place of `generate()`, which means they repurpose all
existing work on the async backend.


## Who can review?

Community members can review the PR once tests pass. Tag
maintainers/contributors who might be interested:
        @hwchase17 (follows his work on #4615)
        @agola11 (async)

---------

Co-authored-by: Harrison Chase <hw.chase.17@gmail.com>

eyurtsev commented on langchain/client/langchain.py:
```diff
@@ -158,8 +164,8 @@ def upload_dataframe(
         description: str,
-        input_keys: List[str],
-        output_keys: List[str],
+        input_keys: Sequence[str],
```
:heart: Immediate approval for PR!

eyurtsev commented on langchain/client/langchain.py:
```diff
@@ -268,7 +271,7 @@ def list_sessions(self) -> Iterator[TracerSession]:
         yield from [TracerSession(**session) for session in response.json()]
 
-    def create_dataset(self, dataset_name: str, description: str) -> Dataset:
+    def create_dataset(self, dataset_name: str, description: Optional[str]) -> Dataset:
```
```suggestion
    def create_dataset(self, dataset_name: str, *, description: Optional[str] = None) -> Dataset:
```

eyurtsev commented on langchain/client/langchain.py:
```diff
@@ -383,6 +386,92 @@ def list_examples(
 
+    def update_example(
+        self,
+        example_id: str,
```
```suggestion
        example_id: str,
        *,
```

eyurtsev commented on langchain/client/langchain.py:
```diff
@@ -383,6 +386,92 @@ def list_examples(
+    def update_example(
+        self,
+        example_id: str,
+        inputs: Optional[Dict[str, Any]] = None,
```
```suggestion
        inputs: Optional[Mapping[str, Any]] = None,
```

eyurtsev commented on langchain/client/langchain.py:
```diff
@@ -383,6 +386,92 @@ def list_examples(
+        self,
+        example_id: str,
+        inputs: Optional[Dict[str, Any]] = None,
+        outputs: Optional[Dict[str, Any]] = None,
```
```suggestion
        outputs: Optional[Mapping[str, Any]] = None,
```

eyurtsev commented on langchain/client/langchain.py:
```diff
@@ -383,6 +386,92 @@ def list_examples(
+        raise_for_status_with_text(response)
+        return Feedback(**feedback.dict())
+
+    @retry(stop=stop_after_attempt(3), wait=wait_fixed(0.5))
```
Are we able to specify which errors are retried? i.e., 5xx and a few selected 4xx

vowelparrot replied:
Ya - I think I'll update that in a subsequent PR though because our error raising isn't super specific

eyurtsev commented on langchain/client/models.py:
```diff
@@ -31,12 +31,20 @@ class Example(ExampleBase):
     runs: List[Run] = Field(default_factory=list)
 
 
+class ExampleUpdate(BaseModel):
```
1) Should we be setting frozen in the config?

vowelparrot replied:
Ya

eyurtsev commented on langchain/client/models.py:
```diff
@@ -31,12 +31,20 @@ class Example(ExampleBase):
+    """Update class for Example."""
+
+    dataset_id: Optional[UUID] = None
+    inputs: Optional[Dict[str, Any]] = None
```
```suggestion
    inputs: Optional[Mapping[str, Any]] = None
```

eyurtsev commented on langchain/client/models.py:
```diff
@@ -31,12 +31,20 @@ class Example(ExampleBase):
+
+    dataset_id: Optional[UUID] = None
+    inputs: Optional[Dict[str, Any]] = None
+    outputs: Optional[Dict[str, Any]] = None
```
```suggestion
    outputs: Optional[Mapping[str, Any]] = None
```

eyurtsev commented on langchain/client/models.py:
```diff
@@ -97,3 +105,57 @@ def validate_time_range(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+
+    created_at: datetime = Field(default_factory=datetime.utcnow)
+    """The time the feedback was created."""
+    modified_at: datetime = Field(default_factory=datetime.utcnow)
```
style: i like `updated_at` better (sounds better to me when contrasted with `created_at`)

vowelparrot replied:
Ya that may be, but we already use modified_at on other resources and I would rather be consistent with that

eyurtsev commented on langchain/client/models.py:
```diff
@@ -97,3 +105,57 @@ def validate_time_range(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+    """The associated run ID this feedback is logged for."""
+    metric_name: str
+    """The feedback metric name or type."""
+    metric_value: Optional[Union[float, str]] = None
```
Should ints / bools be supported?

vowelparrot replied:
Ya they should!

eyurtsev commented on langchain/client/models.py:
```diff
@@ -97,3 +105,57 @@ def validate_time_range(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+    """The feedback metric name or type."""
+    metric_value: Optional[Union[float, str]] = None
+    """Score to assign the run."""
+    feedback_source: Optional[Union[APIFeedbackSource, Dict[str, Any]]] = None
```
```suggestion
    feedback_source: Optional[Union[APIFeedbackSource, Mapping[str, Any]]] = None
```

eyurtsev commented on langchain/client/models.py:
```diff
@@ -97,3 +105,57 @@ def validate_time_range(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+class ListFeedbackQueryParams(BaseModel):
+    """Query Params for listing feedbacks."""
+
+    run: Optional[List[UUID]] = None
```
```suggestion
    run: Optional[Sequence[UUID]] = None
```

eyurtsev reviewed (commented):
Have to run to sleep! Sorry didn't manage to get through the actual PR logic only scanned briefly. (Looking good from what I can tell though.)

mbchang committed:
fix: fix current_time=Now bug for aadd_documents in TimeWeightedRetriever (#5155)

# Same as PR #5045, but for async

<!--
Thank you for contributing to LangChain! Your PR will appear in our next
release under the title you set. Please make sure it highlights your
valuable contribution.

Replace this with a description of the change, the issue it fixes (if
applicable), and relevant context. List any dependencies required for
this change.

After you're done, someone will review your PR. They may suggest
improvements. If no one reviews your PR within a few days, feel free to
@-mention the same people again, as notifications can get lost.
-->

<!-- Remove if not applicable -->

Fixes #4825 

I had forgotten to update the asynchronous counterpart `aadd_documents`
with the bug fix from PR #5045, so this PR also fixes `aadd_documents`
too.

## Who can review?

Community members can review the PR once tests pass. Tag
maintainers/contributors who might be interested:
@dev2049

<!-- For a quicker response, figure out the right person to tag with @

        @hwchase17 - project lead

        Tracing / Callbacks
        - @agola11

        Async
        - @agola11

        DataLoaders
        - @eyurtsev

        Models
        - @hwchase17
        - @agola11

        Agents / Tools / Toolkits
        - @vowelparrot
        
        VectorStores / Retrievers / Memory
        - @dev2049
        
 -->

DanQuin committed:
Docs: updated getting_started.md (#5151)

# Docs: updated getting_started.md

Just accommodating some unnecessary spaces in the example of "pass few
shot examples to a prompt template".

@vowelparrot

DanQuin committed:
Clarification of the reference to the "get_text_legth" function in ge… (#5154)

# Clarification of the reference to the "get_text_legth" function in
getting_started.md

Reference to the function "get_text_legth" in the documentation did not
make sense. Comment added for clarification.

@hwchase17

vowelparrot committed:
Update langchain/client/langchain.py

Co-authored-by: Eugene Yurtsev <eyurtsev@gmail.com>

vowelparrot committed:
Update langchain/client/langchain.py

Co-authored-by: Eugene Yurtsev <eyurtsev@gmail.com>

vowelparrot committed:
Update langchain/client/langchain.py

Co-authored-by: Eugene Yurtsev <eyurtsev@gmail.com>

vowelparrot committed:
Update langchain/client/models.py

Co-authored-by: Eugene Yurtsev <eyurtsev@gmail.com>

vowelparrot committed:
Update langchain/client/models.py

Co-authored-by: Eugene Yurtsev <eyurtsev@gmail.com>

vowelparrot committed:
type

vowelparrot committed:
Update langchain/client/langchain.py

Co-authored-by: Eugene Yurtsev <eyurtsev@gmail.com>

vowelparrot committed:
Update langchain/client/models.py

Co-authored-by: Eugene Yurtsev <eyurtsev@gmail.com>

vowelparrot commented:
Accidentally slip-fingered a reverse rebase and messed up this PR. Going to continue in #5166
